pkg/entry/mongo/document.go

```diff
@@ -1,163 +1,254 @@
+    }
+
+    ts, err := parse.ExtractStringValue(record, TimeKey)
+    if err != nil {
+        if !errors.Is(err, &parse.ErrKeyNotFound{
+            LookingFor: TimeKey,
+        }) {
+            return fmt.Errorf("parse %s: %w", TimeKey, err)
+        }
+
+        logger.Debug("Key not found", map[string]interface{}{
+            "error": err,
+        })
+    } else {
+        d.Time = ts
+    }
+
+    d.ProjectId, err = parse.ExtractStringValue(record, ProjectIDKey)
+    if err != nil {
+        return fmt.Errorf("parse %s: %w", ProjectIDKey, err)
+    }
+
+    d.Customer, err = parse.ExtractStringValue(record, CustomerKey)
+    if err != nil {
+        return fmt.Errorf("parse %s: %w", CustomerKey, err)
+    }
+
+    d.PlatformId, err = parse.ExtractStringValue(record, PlatformIDKey)
+    if err != nil {
+        return fmt.Errorf("parse %s: %w", PlatformIDKey, err)
+    }
+
+    return nil
 }
 
-func (d *Document) CollectionName() string {
-	return strings.Replace(fmt.Sprintf("%s_%s_%s", d.Customer, d.PlatformId, d.ProjectId), "-", "_", -1)
+func (d *LogDocument) generateObjectID() error {
+    logJson, err := json.Marshal(d)
+    if err != nil {
+        return err
+    }
+
+    h64bytes, h32bytes, err := parse.GetHashesFromBytes(logJson)
+    if err != nil {
+        return err
+    }
+
+    id := fmt.Sprintf("%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x",
+        h64bytes[0], h64bytes[1], h64bytes[2], h64bytes[3], h64bytes[4], h64bytes[5], h64bytes[6], h64bytes[7],
+        h32bytes[0], h32bytes[1], h32bytes[2], h32bytes[3],
+    )
+    d.Id = bson.ObjectIdHex(id)
+    return nil
+}
+
+func (d *LogDocument) CollectionName() string {
+    return strings.Replace(fmt.Sprintf("%s_%s_%s", d.Customer, d.PlatformId, d.ProjectId), "-", "_", -1)
+}
+
+func (d *JobLogDocument) SaveTo(collection *mgo.Collection) error {
+    if _, err := collection.UpsertId(d.Id, d); err != nil {
+        return fmt.Errorf("upsert %s: %w", d.Id, err)
+    }
+
+    indexes := []string{"job_execution_id", "time"}
+
+    if err := collection.EnsureIndexKey(indexes...); err != nil {
+        return fmt.Errorf("ensure indexes %v: %w", indexes, err)
+    }
+
+    return nil
 }
 
-func (d *Document) SaveTo(collection *mgo.Collection) error {
-	if _, err := collection.UpsertId(d.Id, d); err != nil {
-		return fmt.Errorf("upsert %s: %w", d.Id, err)
-	}
+func (d *AppLogDocument) SaveTo(collection *mgo.Collection) error {
+    if _, err := collection.UpsertId(d.Id, d); err != nil {
+        return fmt.Errorf("upsert %s: %w", d.Id, err)
+    }
 
-	indexes := []string{"job_execution_id", "time"}
+    indexes := []string{"app_execution_id", "container_id", "time"}
 
-	if err := collection.EnsureIndexKey(indexes...); err != nil {
-		return fmt.Errorf("ensure indexes %v: %w", indexes, err)
-	}
+    if err := collection.EnsureIndexKey(indexes...); err != nil {
+        return fmt.Errorf("ensure indexes %v: %w", indexes, err)
+    }
 
-	return nil
+    return nil
```

Comment:
Why did you replace tabulation with spaces?

Reply:
Typically to avoid 
```
ContainerIDKey	  = "container_id"
AppExecutionIDKey = "app_execution_id"
AppIDKey 		  = "app_id"
```
I prefer directly configure the IDE for use only one type of space character to avoid mixing them.

Reply:
Fuck bad IDE!

You should either:

1. Configure your IDE to use space for assignment alignment
2. Use a good IDE :-p (I suggest Visual Studio Code)
3. Or update all files of the repository to use spaces as indentation so we have same indentation configuration for all. (This would be really painful to review)

Reply:
Done by reverting. But It baffles me that we used tabulation to start a line but spaces to align the properties/types